client/network/src/lib.rs

```diff
@@ -321,3 +330,99 @@ pub struct NetworkStatus<B: BlockT> {
 	/// The total number of bytes sent.
 	pub total_bytes_outbound: u64,
 }
+
+/// Provides access to preimported blocks.
+/// A preimported block is one that has been announced, checked and verified
+/// but not yet imported into the chain's backend.
+pub struct VerifiedBlocks<B: BlockT> {
+	client: Arc<dyn crate::chain::Client<B>>,
+	blocks: RwLock<HashMap<B::Hash, (B, bool)>>,
+}
+
+impl<B: BlockT> VerifiedBlocks<B> {
+	/// Returns VerifiedBlocks
+	pub fn new(client: Arc<dyn crate::chain::Client<B>>) -> Self {
+		Self {
+			client,
+			blocks: RwLock::new(HashMap::new()),
+		}
+	}
+
+	/// Register a new block which has been announced
+	/// and downloaded from peers.
+	fn extend(&self, blocks: Vec<B>) {
+		self.blocks.write().extend(blocks.into_iter().map(|b| (b.hash(), (b, false))));
+	}
+
+	/// Fetch the header of a block.
+	/// This will try to fetch the imported block from the client's backend.
```

Comment:
Why do we first try to check the client? Checking the preimport cache should be much faster.